xds/internal/xdsclient/authority_test.go

```diff
@@ -0,0 +1,344 @@
+		t.Fatal(err)
+	}
+
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{e2e.DefaultClientListener(resourceName, "new-rds-resource")},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Waiting for authority to receive the update before verifying state.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out before watcher received the update.")
+	case <-watcher.updateCh:
+	}
+
+	if err := compareWStateAndWTimer(a, listenerResourceType, resourceName, watchStateReceived); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// This tests the resource's watch state transition when the ADS stream is closed
+// by the management server. The test calls the `watchResource` api to register
+// a watch for a resource, stops the management server, and verifies the resource's
+// watch state transitions from `watchStateRequested` to `watchStateStarted` so
+// that the watch can be restarted later.
+func (s) TestResourceStateTransitionsFromRequestedToStartedOnError(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	onStreamRequestDoneCh := make(chan struct{})
+	mgmtServerOpts := mgmtServerOptWithOnRequestDoneCh(onStreamRequestDoneCh)
+	a, mgmtServer, _ := setupAuthorityWithMgmtServer(ctx, t, mgmtServerOpts, defaultTestTimeout)
+	defer a.close()
+
+	resourceName := "xdsclient-test-lds-resource"
+	watcher := newTestResourceWatcher()
+	cancelResource := a.watchResource(listenerResourceType, resourceName, watcher)
+	defer cancelResource()
+
+	// Waiting for mgmt server to recv request.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out before mgmt server got the request.")
+	case <-onStreamRequestDoneCh:
+	}
+	if err := compareWStateAndWTimer(a, listenerResourceType, resourceName, watchStateRequested); err != nil {
+		t.Fatal(err)
+	}
+
+	mgmtServer.Stop()
+
+	// Waiting for watcher to receive the update.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out before verifying error propagation.")
+	case err := <-watcher.errorCh:
+		if err == nil {
+			t.Fatal("got err == nil. Want: err from stream connection reset by peer")
+		}
+	}
+	if err := compareWStateAndWTimer(a, listenerResourceType, resourceName, watchStateStarted); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// This tests the case where ADS stream breaks after a successfully receiving a
+// message on the stream. In this case, we do not want the error to be propagated
+// to the watchers. verifies that in the case where the ADS stream breaks after successfully
+// receiving a message on the stream. In this case we want to ignore propagating
+// connection error to the watcher. But since the mgmt server stops in this test,
+// the watcher would be updated with a connection error for the subsequent attempt
+// to create an ADS stream.
```

Comment:
This comment could be better since the test is actually doing more.

// This tests the case where the ADS stream breaks after successfully receiving a
// message on the stream. The test performs the following:
// - configures the management server with resourceA.
// - registers a watch for resourceA and verifies that the watcher's update callback is invoked.
// - registers a watch for resourceB and verifies that the watcher's update callback is not invoked. This is because the management server does not contain resourceB.
// - stop the management server.
// - verifies that the error propagated to the watcher is a connection error. This happens when the authority attempts to create a new stream.

Need not exactly be like this. But I guess, you get the point.